### PR TITLE
Disable failing patron authentication (DB Migration Required)

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description
The LibraryAuthenticator skips authentication mechanisms based on their status
    
There is also a random sampling built into the disable logic
so the disabled integrations get sampled during the timeout window
Now instead of raising an error the authenticated_patron method will return a problem detail
This is so the error is written to the DB and not rolled back at the end of the request

## Motivation and Context
Over the last several weeks, we have seen a number of patron authentication mechanisms fail (the remote ILS goes offline). Because these connections take some time to fail, they cause a backup of requests on the webserver, causing other library on the same CM not to be able to authenticate.

[Notion](https://www.notion.so/lyrasis/Disable-patron-authentication-mechanism-after-repeated-failures-459919ba61034183be2221ee762d94a5)

## How Has This Been Tested?
Manually Set and unset the status while hitting an authenticated endpoint
New unit tests were written

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/authenticator.py
Comment (RishiDiwanTT) [inline api/authenticator.py:891]: Randomly chosen sample rate.

File: api/sip/client.py
Comment (RishiDiwanTT) [inline api/sip/client.py:357]: In case the credentials provided are incorrect, it should not contribute to the failure counts of the auth mechanism, hence this exception type was changed to differentiate the login failure vs API failure

**Review by dbernstein — APPROVED** [review]
Looks good to me. Nice work, @RishiDiwanTT.

File: core/jobs/update_integration_status.py
Comment (RishiDiwanTT) [inline core/jobs/update_integration_status.py:17]: I forgot to draw attention to these numbers, they were chosen because they are numbers, and for no other reason.
Should they be changed to something with more rationale behind them?
@dbernstein @jonathangreen

File: core/model/configuration.py
Comment (tdilauro) [inline core/model/configuration.py:150]: Any reason not to just make this an enum?

**Review by tdilauro — CHANGES_REQUESTED** [review]
This is looking pretty good!

I have a couple concerns, which are discussed in the comments below, and one minor general general suggestion, which is to end sentences in comments and docstrings with a period/full stop.

File: api/sip/client.py
Comment (RishiDiwanTT) [reply api/sip/client.py:357]: A failed login would usually mean the API server is running just fine and the User has incorrectly input their credentials. This should not impact other Users on the system who would be using the functioning API for authentication.

The response being logged and returned in the Exception is a json value, we cannot reliably depend on it string format to differentiate the types. I also believe having a different Exception type altogether is more explicit and future-proof as opposed to differentiating based on the string messages within them.

File: core/model/configuration.py
Comment (RishiDiwanTT) [reply core/model/configuration.py]: Laziness. This has been fixed. Not in me, in the code.

Comment (jonathangreen) [reply core/model/configuration.py]: I was reading through my GitHub notifications this morning while I was eating breakfast, and this comment gave me a good laugh 🤣

File: core/jobs/update_integration_status.py
Comment (dbernstein) [reply core/jobs/update_integration_status.py:17]: @RishiDiwanTT :  I like numbers.  Numbers are good.  ;) It seems like as good a guess as any.  I think 10 errors in 1/2  hour is a good place to start.

**Comment by RishiDiwanTT** [comment]
@tdilauro I'll wait for your approval before merging.

File: api/sip/client.py
Comment (tdilauro) [reply api/sip/client.py:357]: I was focused in on SIP2 service, where login is distinct from patron lookup. In that (SIP2) case, login failure would, in a practical sense, be tantamount to service unavailability. But that is not the case for all of our integrations.

File: api/authenticator.py
Comment (tdilauro) [reply api/authenticator.py]: I think this one might've been accidentally resolved, so unresolving. ☝🏽

Comment (RishiDiwanTT) [reply api/authenticator.py]: I removed this logic in this [commit](https://github.com/ThePalaceProject/circulation/pull/980/commits/d456f194e94fe86fc77828f81e0a883d50912ca7) hence marked it as resolved

**Review by tdilauro — APPROVED** [review]
I think this is ready to go! 🏁

**Pull request merged**
